pkg/tfgen/generate.go

```diff
@@ -115,8 +112,120 @@ func (l Language) shouldConvertExamples() bool {
 	return false
 }
 
+func dirToBytesMap(fs afero.Fs, dir string) (map[string][]byte, error) {
+	result := make(map[string][]byte)
+	err := afero.Walk(fs, dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		relPath, err := filepath.Rel(dir, path)
+		if err != nil {
+			return err
+		}
+		if relPath == "." {
+			return nil
+		}
+		content, err := afero.ReadFile(fs, path)
+		if err != nil {
+			return err
+		}
+		result[relPath] = content
+		return nil
+	})
+	return result, err
+}
+
+func writeBytesMapToDir(fs afero.Fs, dir string, files map[string][]byte) error {
+	err := fs.MkdirAll(dir, 0o755)
+	if err != nil {
+		return pkgerrors.Wrap(err, "failed to create dir")
+	}
+	for name, content := range files {
+		srcDir := filepath.Dir(name)
+		if srcDir != "." {
+			err = fs.MkdirAll(filepath.Join(dir, srcDir), 0o755)
+			if err != nil {
+				return pkgerrors.Wrap(err, "failed to create dir")
+			}
+		}
+		err := afero.WriteFile(fs, filepath.Join(dir, name), content, 0o600)
+		if err != nil {
+			return pkgerrors.Wrap(err, "failed to write file")
+		}
+	}
+	return nil
+}
+
+func runPulumiPackageGenSDK(l Language, pkg *pschema.Package, extraFiles map[string][]byte) (map[string][]byte, error) {
+	var err error
+
+	fs := afero.NewOsFs()
+	outDir, err := afero.TempDir(fs, "", "pulumi-package-gen-sdk")
+	if err != nil {
+		return nil, pkgerrors.Wrap(err, "failed to create temp dir")
+	}
+	defer func() {
+		rmErr := fs.RemoveAll(outDir)
+		if rmErr != nil {
+			err = multierror.Append(err, rmErr)
+		}
+	}()
+
+	args := []string{"package", "gen-sdk", "--language", string(l), "--out", outDir}
+
+	if len(extraFiles) > 0 {
+		overlayDir, err := afero.TempDir(fs, "", "pulumi-package-gen-sdk-overlays")
+		if err != nil {
+			return nil, pkgerrors.Wrap(err, "failed to create temp dir")
+		}
+		defer func() {
+			rmErr := fs.RemoveAll(overlayDir)
+			if rmErr != nil {
+				err = multierror.Append(err, rmErr)
+			}
+		}()
+		dest := filepath.Join(overlayDir, string(l))
+		err = writeBytesMapToDir(fs, dest, extraFiles)
+		if err != nil {
+			return nil, pkgerrors.Wrap(err, "failed to write overlay files")
+		}
+
+		args = append(args, "--overlays", overlayDir)
+	}
+
+	schemaDir, err := afero.TempDir(fs, "", "schema")
+	if err != nil {
+		return nil, pkgerrors.Wrap(err, "failed to create temp dir")
+	}
+	schemaFile := filepath.Join(schemaDir, "schema.json")
+	schemaBytes, err := pkg.MarshalJSON()
+	if err != nil {
+		return nil, pkgerrors.Wrap(err, "failed to marshal schema")
+	}
+	err = afero.WriteFile(fs, schemaFile, schemaBytes, 0o600)
+	if err != nil {
+		return nil, pkgerrors.Wrap(err, "failed to write schema")
+	}
+
+	args = append(args, schemaFile)
+
+	cmd := exec.Command("pulumi", args...)
```

Comment:
Worth calling out in release notes that `pulumi` needs to be pinned in the builds for most users, or else the builds become unstable with respect to generated SDKs. Similar to https://github.com/pulumi/pulumi-terraform-bridge/issues/2037 but scope here seems smaller because language plugins are not independently versioned. Perhaps we can consider rolling out 2037 to make that the default in the same release.

Reply:
Also confirming - `pulumi` is pinned for tests in this repo right? That would be needed for any golden tests around this.